src/hist/plot.py

```diff
@@ -214,185 +263,402 @@ def plot2d_full(
+    ub_kwargs: Dict[str, Any],
+) -> FitResultArtists:
+    """
+    Plot fit of model to histogram data
+    """
+    x_values = _hist.axes[0].centers
+    variances = _hist.variances()
+    if variances is None:
+        raise RuntimeError(
+            "Cannot compute from a variance-less histogram, try a Weight storage"
+        )
+    hist_uncert = np.sqrt(variances)
+
+    errorbars = ax.errorbar(x_values, _hist.values(), hist_uncert, **eb_kwargs)
+
+    # Ensure zorder draws data points above model
+    line_zorder = errorbars[0].get_zorder() - 1
+    (line,) = ax.plot(x_values, model_values, **fp_kwargs, zorder=line_zorder)
+
+    # Uncertainty band for fitted function
+    # TODO: Probably set a better default color than the fit line color
+    ub_kwargs.setdefault("color", line.get_color())
+    uncertainty_band = ax.fill_between(
+        x_values,
+        model_values - model_uncert,
+        model_values + model_uncert,
+        **ub_kwargs,
+    )
+
+    return FitResultArtists(line, errorbars, uncertainty_band)
+
+
+def plot_ratio(
+    _hist: hist.BaseHist,
+    ratio: np.ndarray,
+    ratio_uncert: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    **kwargs: Any,
+) -> RatioArtists:
+    """
+    Plot a ratio plot on the given axes
+    """
+    x_values = _hist.axes[0].centers
+    left_edge = _hist.axes.edges[0][0]
+    right_edge = _hist.axes.edges[-1][-1]
+
+    # Set 0 and inf to nan to hide during plotting
+    ratio[ratio == 0] = np.nan
+    ratio[np.isinf(ratio)] = np.nan
+
+    central_value = kwargs.pop("central_value", 1.0)
+    central_value_artist = ax.axhline(
+        central_value, color="black", linestyle="dashed", linewidth=1.0
+    )
+
+    uncert_draw_type = kwargs.pop("uncert_draw_type", "line")
+    if uncert_draw_type == "line":
+        errorbar_artists = ax.errorbar(
+            x_values,
+            ratio,
+            yerr=ratio_uncert,
+            color="black",
+            marker="o",
+            linestyle="none",
+        )
+        axis_artists = RatioErrorbarArtists(central_value_artist, errorbar_artists)
+    elif uncert_draw_type == "bar":
+        bar_width = (right_edge - left_edge) / len(ratio)
+
+        bar_top = ratio + ratio_uncert[1]
+        bar_bottom = ratio - ratio_uncert[0]
+        # bottom can't be nan
+        bar_bottom[np.isnan(bar_bottom)] = 0
+        bar_height = bar_top - bar_bottom
+
+        _ratio_points = ax.scatter(x_values, ratio, color="black")
+
+        # Ensure zorder draws data points above uncertainty bars
+        bar_zorder = _ratio_points.get_zorder() - 1
+        bar_artists = ax.bar(
+            x_values,
+            height=bar_height,
+            width=bar_width,
+            bottom=bar_bottom,
+            fill=False,
+            linewidth=0,
+            edgecolor="gray",
+            hatch=3 * "/",
+            zorder=bar_zorder,
+        )
+        axis_artists = RatioBarArtists(central_value_artist, _ratio_points, bar_artists)
```

Comment:
@henryiii I'm getting errors (here and few other places for similar reasons) that 

```
src/hist/plot.py:423: error: Incompatible types in assignment (expression has type "RatioBarArtists", variable has type "RatioErrorbarArtists")
```

because in the control flow above for the other branch it is set as

```python
axis_artists = RatioErrorbarArtists(central_value_artist, errorbar_artists)
```

If I try and type both explicitly

```python
        axis_artists: RatioErrorbarArtists = RatioErrorbarArtists(
            central_value_artist, errorbar_artists
        )
        # ....
        axis_artists: RatioBarArtists = RatioBarArtists(
            central_value_artist, _ratio_points, bar_artists
        )
```

then I get 

```
src/hist/plot.py:425: error: Name 'axis_artists' already defined on line 398
```

Am I missing the obvious with how to deal with the control flow here? Or does this require type checking to be skipped at these lines?

Reply:
```python
axis_artists: Union[RatioBarArtists, RatioErrorbarArtists]

axis_artists = RatioErrorbarArtists(
    central_value_artist, errorbar_artists
)
# ....
axis_artists = RatioBarArtists(
    central_value_artist, _ratio_points, bar_artists
)
```

Reply:
Thanks.